Dependency exclusions on extensions are ignored

essobedo opened the pull request:
Fixes #14485 

## Motivation

We have no way to exclude a dependency from a Quarkus extension.

## Modifications:

* Adds the exclusion of the original `DependencyNode` when replacing it with it deployment counter part
* Adds the ability to exclude dependencies from `TsArtifact`

essobedo committed:
Dependency exclusions on extensions are ignored

aloubyansky commented:
Thanks a lot for the PR @essobedo This is definitely a bug we need to fix. The changes look pretty good. I have a few remarks though.

These changes apply the exclusions configured for the runtime dependency to the deployment dependency, e.g. in the contributed test none of the exclusions except '3' appear to be runtime dependencies. That doesn't look nice. The advantage of doing it this way is it makes it really easy to exclude deps from the deployment dependency. The disadvantage is it really forces the exclusions on the deployment part, making it impossible to apply exclusions to only the runtime dependency.
If we propagate these exclusions to the runtime dependency only though (which is a direct dep of the deployment one) to be able to apply exclusions to the deployment dep (that aren't contributed by the runtime dep, i.e. anything but '3' in the contributed testcase) the deployment dependency would have to be added to the `dependencyManagement` element with the necessary exclusions. That would make configuring exclusions for the deployment dep (which is probably an edge case anyway) more tricky but on the other hand that'd be more precise.

However, there is another use-case. Excluding a transitive extension dependency. With these changes we should definitely add a testcase for this too. So, with your current impl it should just work. But it will be tricky with the alternative I mentioned above.

aloubyansky commented:
BTW, by the transitive extension dependency I mean a use-case when extension A depends on extension B and we want to exclude B.

aloubyansky commented:
@essobedo would you mind adding another test that excludes a transitive dependency on an extension?

essobedo commented:
> @essobedo would you mind adding another test that excludes a transitive dependency on an extension?

@aloubyansky ok no problem, I can add it

essobedo commented:
@aloubyansky I improved the test to have the dependencies as runtime dependencies and I added the case where the application depends on an extension C which itself depends on B and excludes B. Is it what you had in mind?

aloubyansky commented:
Looks like it @essobedo You could have added another test class for testing the extension exclusion, to not make the first test more complicated. I'll review it again on Monday. Thanks again!

essobedo commented:
@aloubyansky ok I will split the test in 2 for clarity

essobedo commented:
@aloubyansky done, I also squashed and rebased because I prefer keeping one commit hoping that it's not an issue for you. Thx in advance for the review 🙏

aloubyansky reviewed (approved):
Great work @essobedo Thanks!
